src/main/cpp/src/multiply.cu

```diff
@@ -16,99 +16,338 @@
+  // collect null mask and set
+  auto [null_mask, null_count] =
+    cudf::detail::valid_if(validity.begin(), validity.end(), cuda::std::identity{}, stream, mr);
+  if (null_count > 0) { result->set_null_mask(std::move(null_mask), null_count, stream); }
+
+  return result;
+}
+
+/**
+ * @brief Operator for multiply(cv, cv).
+ */
+struct dispatch_multiply {
+  // left operand can be either a column_view or a scalar
+  cudf::column_view const* left_cv;
+  cudf::scalar const* left_scalar;
+
+  // right operand can be either a column_view or a scalar
+  cudf::column_view const* right_cv;
+  cudf::scalar const* right_scalar;
+
+  template <typename T, CUDF_ENABLE_IF(is_basic_spark_numeric<T>())>
+  std::unique_ptr<cudf::column> operator()(cudf::data_type type,
+                                           cudf::size_type num_rows,
+                                           bool check_overflow,
+                                           rmm::cuda_stream_view stream,
+                                           rmm::device_async_resource_ref mr) const
+  {
+    if (left_cv != nullptr && right_cv != nullptr) {
+      auto const left_cdv  = cudf::column_device_view::create(*left_cv, stream);
+      auto const right_cdv = cudf::column_device_view::create(*right_cv, stream);
+      if (left_cv->has_nulls()) {
+        if (right_cv->has_nulls()) {
+          auto const left_accessor  = cudf::detail::make_pair_iterator<T, true>(*left_cdv);
+          auto const right_accessor = cudf::detail::make_pair_iterator<T, true>(*right_cdv);
+          return multiply_impl<T, decltype(left_accessor), decltype(right_accessor)>(
+            type, num_rows, left_accessor, right_accessor, check_overflow, stream, mr);
+        } else {
+          auto const left_accessor  = cudf::detail::make_pair_iterator<T, true>(*left_cdv);
+          auto const right_accessor = cudf::detail::make_pair_iterator<T, false>(*right_cdv);
+          return multiply_impl<T, decltype(left_accessor), decltype(right_accessor)>(
+            type, num_rows, left_accessor, right_accessor, check_overflow, stream, mr);
+        }
+      } else {
+        if (right_cv->has_nulls()) {
+          auto const left_accessor  = cudf::detail::make_pair_iterator<T, false>(*left_cdv);
+          auto const right_accessor = cudf::detail::make_pair_iterator<T, true>(*right_cdv);
+          return multiply_impl<T, decltype(left_accessor), decltype(right_accessor)>(
+            type, num_rows, left_accessor, right_accessor, check_overflow, stream, mr);
+        } else {
+          auto const left_accessor  = cudf::detail::make_pair_iterator<T, false>(*left_cdv);
+          auto const right_accessor = cudf::detail::make_pair_iterator<T, false>(*right_cdv);
+          return multiply_impl<T, decltype(left_accessor), decltype(right_accessor)>(
+            type, num_rows, left_accessor, right_accessor, check_overflow, stream, mr);
+        }
+      }
+    } else if (left_cv != nullptr && right_scalar != nullptr) {
```

Comment:
In addition, I believe that we can shorten this function can be shorten significantly more using `std::visit`. However, it is optionial as I don't want to push too hard for it 😄